Source/WebKit/WebProcess/WebPage/WebPage.cpp

```diff
@@ -10276,6 +10276,37 @@ void WebPage::setPresentingApplicationAuditToken(CoreIPCAuditToken&& presentingA
 }
 #endif
 
+#if ENABLE(CONTENT_EXTENSIONS)
+static String generateResourceMonitorErrorHTML()
+{
+#if PLATFORM(COCOA) && HAVE(LOCAL_FRAME_ADDITIONS)
+    return generateResourceMonitorErrorHTMLForCocoa();
+#else
+    return WEB_UI_STRING("This frame is hidden for using too many system resources.", "Description HTML for frame unloaded by ResourceMonitor");
+#endif
+}
+
+void WebPage::didExeedFrameNetworkUsage(WebCore::FrameIdentifier frameID)
+{
+    RefPtr frame = WebProcess::singleton().webFrame(frameID);
+    if (!frame)
+        return;
+
+    RefPtr coreLocalFrame = frame->coreLocalFrame();
+    if (!coreLocalFrame)
+        return;
+
+    ASSERT(!coreLocalFrame->isMainFrame());
+
+    WEBPAGE_RELEASE_LOG(Loading, "didExeedFrameNetworkUsage: Unloading frame due to exceeding threshold.");
+
+    coreLocalFrame->showResourceMonitoringError(generateResourceMonitorErrorHTML());
+
+    if (RefPtr document = coreLocalFrame->document())
+        document->addConsoleMessage(MessageSource::ContentBlocker, MessageLevel::Error, "Frame was unloaded because its network usage exceeded the limit."_s);
```

Comment:
Should this be moved to `showResourceMonitoringError()`? We want as much as possible of the implementation to be in WebCore so it can potentially be shared with WebKitLegacy in the future.